knowledge_repo/app/auth_providers/oauth2.py

```diff
@@ -0,0 +1,165 @@
+        'auto_refresh_url': 'https://accounts.google.com/o/oauth2/token',
+        'scopes': [
+            'https://www.googleapis.com/auth/userinfo.email',
+            'https://www.googleapis.com/auth/userinfo.profile'
+        ],
+        'user_info_endpoint': 'userinfo',
+        'user_info_mapping': {
+            'identifier': 'email',
+            'name': 'name',
+            'photo_uri': 'picture'
+        }
+    }
+}
+
+
+def _resolve_oauth_config(name, local_namespace, config, *variables):
+    presets = PRESETS.get(name, {})
+    output = []
+    for variable in variables:
+        value = local_namespace.get(variable, None)
+        if value is None:
+            value = config.get('OAUTH_{}_{}'.format(name, variable).upper(), None)
+        if value is None:
+            value = presets.get(variable, None)
+        output.append(value)
+    return output
+
+
+class OAuth2Provider(KnowledgeAuthProvider):
+    _registry_keys = list(PRESETS.keys())
+
+    def init(self,
+             authorization_url=None,
+             token_url=None,
+             auto_refresh_url=None,
+             scopes=None,
+             user_info_endpoint=None,
+             client_id=None,
+             client_secret=None,
+             user_info_mapping=None,
+             validate=None):
+
+        (self.base_url,
+         self.authorization_url,
+         self.token_url,
+         self.auto_refresh_url,
+         self.scopes,
+         self.user_info_endpoint,
+         self.client_id,
+         self.client_secret,
+         self.user_info_mapping,
+         validate) = _resolve_oauth_config(
+            self.name,
+            locals(),
+            self.app.config,
+            'base_url',
+            'authorization_url',
+            'token_url',
+            'auto_refresh_url',
+            'scopes',
+            'user_info_endpoint',
+            'client_id',
+            'client_secret',
+            'user_info_mapping',
+            'validate'
+        )
+        if validate is not None:
+            self.validate = validate
+
+        redirect_url = self.app.config['SERVER_NAME'] or 'localhost:7000'
+        if self.app.config['APPLICATION_ROOT']:
+            redirect_url = '/'.join((redirect_url, self.app.config['APPLICATION_ROOT']))
+
+        # Import OAuth deps here so we do not have a hard dependency on them
+        from requests_oauthlib import OAuth2Session
+        self.oauth_client = OAuth2Session(
+            client_id=self.client_id,
+            scope=self.scopes,
+            redirect_uri='http://{}/auth/login/{}/authorize'.format(redirect_url, self.name),
+            auto_refresh_url=self.auto_refresh_url
+        )
+
+    def prompt(self):
+        return redirect(self.oauth_client.authorization_url(self.authorization_url)[0])
+
+    def get_endpoint_url(self, endpoint):
+        return urljoin(self.base_url, endpoint)
+
+    def get_user(self):
+        code = request.args.get('code')
```

Comment:
[question] Do we want a default value here? Or are we okay passing in None to `fetch_token`